src/components/mc/cuda/kernel/mc_cuda_reduce.cu

```diff
@@ -0,0 +1,168 @@
+        case UCC_OP_BXOR:                                                      \
+            LAUNCH_KERNEL(BXOR, type, sbuf1, sbuf2, dest, count, s, b, t);     \
+            break;                                                             \
+        default:                                                               \
+            mc_error(&ucc_mc_cuda.super, "int dtype does not support "         \
+                                         "requested reduce op: %d", op);       \
+            return UCC_ERR_NOT_SUPPORTED;                                      \
+        }                                                                      \
+    } while(0)
+
+#define DT_REDUCE_FLOAT(type, op, src1_p, src2_p, dest_p, count, s, b, t) do { \
+        const type *sbuf1 = (const type *)src1_p;                              \
+        const type *sbuf2 = (const type *)src2_p;                              \
+        type *dest = (type *)dest_p;                                           \
+        switch(op) {                                                           \
+        case UCC_OP_MAX:                                                       \
+            LAUNCH_KERNEL(MAX, type, sbuf1, sbuf2, dest, count, s, b, t);      \
+            break;                                                             \
+        case UCC_OP_MIN:                                                       \
+            LAUNCH_KERNEL(MIN, type, sbuf1, sbuf2, dest, count, s, b, t);      \
+            break;                                                             \
+        case UCC_OP_SUM:                                                       \
+            LAUNCH_KERNEL(SUM, type, sbuf1, sbuf2, dest, count, s, b, t);      \
+            break;                                                             \
+        case UCC_OP_PROD:                                                      \
+            LAUNCH_KERNEL(PROD, type, sbuf1, sbuf2, dest, count, s, b, t);     \
+            break;                                                             \
+        default:                                                               \
+            mc_error(&ucc_mc_cuda.super, "float dtype does not support "       \
+                                         "requested reduce op: %d", op);       \
+            return UCC_ERR_NOT_SUPPORTED;                                      \
+        }                                                                      \
+    } while(0)
+
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+ucc_status_t ucc_mc_cuda_reduce(const void *src1, const void *src2, void *dst,
+                                size_t count, ucc_datatype_t dt,
+                                ucc_reduction_op_t op)
+{
+    int th;
+    unsigned long bk;
+    cudaStream_t stream;
+
+    stream = ucc_mc_cuda.stream;
+    th = MC_CUDA_CONFIG->reduce_num_threads;
+    bk = (count + th - 1)/th;
```

Comment:
just for my understanding: if user does not set required number of blocks, are we trying to get as many blocks as it is needed to have 1 op per thread ? Ie, total number of threads == count ?

Reply:
correct